lib/mongoid/criteria/queryable/selector.rb

```diff
@@ -21,7 +21,11 @@ def merge!(other)
           other.each_pair do |key, value|
             if value.is_a?(Hash) && self[key.to_s].is_a?(Hash)
               value = self[key.to_s].merge(value) do |_key, old_val, new_val|
-                multi_value?(_key) ? (old_val + new_val).uniq : new_val
+                if multi_value?(_key)
+                  _key == '$in' ? new_val & old_val : old_val + new_val
```

Comment:
This line should be:

```
_key == '$in' ? new_val & old_val : old_val | new_val
```

(`|` operator performs `uniq`). Otherwise you break the `$nin` case.